atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java

```diff
@@ -263,11 +266,52 @@ public InetSocketAddress getRandomHostForKey(byte[] key) {
         if (liveOwnerHosts.isEmpty()) {
             log.warn("Perf / cluster stability issue. Token aware query routing has failed because there are no known " +
                     "live hosts that claim ownership of the given range. Falling back to choosing a random live node. " +
-                    "For debugging, our current ring view is: {} and our current host blacklist is {}", tokenMap, blacklistedHosts);
+                    "Current state logged at DEBUG");
+            log.debug("Current ring view is: {} and our current host blacklist is {}", tokenMap, blacklistedHosts);
             return getRandomGoodHost().getHost();
         } else {
-            return currentPools.get(ImmutableList.copyOf(liveOwnerHosts).get(ThreadLocalRandom.current().nextInt(liveOwnerHosts.size()))).getHost();
+            return getRandomHostByActiveConnections(Maps.filterKeys(currentPools, liveOwnerHosts::contains));
+        }
+    }
+
+    private static InetSocketAddress getRandomHostByActiveConnections(Map<InetSocketAddress, CassandraClientPoolingContainer> pools) {
+        TreeMap<Integer, InetSocketAddress> weightedHosts = buildHostsWeightedByActiveConnections(pools);
+        return weightedHosts.ceilingEntry(ThreadLocalRandom.current().nextInt(weightedHosts.lastKey())).getValue();
```

Comment:
This is wrong. I'm just going to factor the TreeMap into a separate class so I can test it and document it more clearly